# Homework 2: Pandas, MatPlotLib, and SciKit-Learn
**Created by:**&emsp;**Jacob Norman**  
**Date:**&emsp;&emsp;&emsp;&nbsp;&nbsp;&nbsp;&nbsp;**3/8/2019**
## Overview
This homework assignment covers the remaining content from Jake Vanderplas' text *Python Data Science Handbook*. This includes ``pandas``, ``matplotlib`` (and ``seaborn``), and ``sklearn``. I decided to group these three chapters together because they are so tightly linked anyway. Group by analysis with ``pandas`` is often supplemented with plotting in ``matplotlib`` or ``seaborn``. This is often known as *exploratory data analysis*, or EDA. The first part of this assignment will involve some EDA on a baseball dataset. Afterwords, I will use this analysis as a basis for a machine learning model with ``sklearn``.
## About the Data
The data used for this assignment is pitch-by-pitch data for Justin Verlander for the entire 2017 MLB season, including the postseason. Perhaps 2011, when he won the AL MVP, AL Cy Young, and AL pitching Triple Crown (W, ERA, Strikeouts), would have been a better year to analyze; however, the data was not as complete in this year. You may remember 2017 as the year the Detroit Tigers traded JV to the Houston Astros. In return the Tigers got some solid prospects, Jake Rogers (C), Daz Cameron (OF), and Franklin Perez (RHP). JV's hefty contract was a principal factor in the deal, hence why he was traded after the waiver-free deadline. To seal the deal, the Tigers had to pay \$8 million of JV's contract in 2018 and 2019. However, having that contract off the Tiger's payroll has freed them up and allowed them to accelerate their rebuild. The deal worked out for JV, as he finally became a World Series Champion. He was a key component in the Astro's pitching rotation, arguably being the reason they won the World Series. 

PITCHf/x is a relatively new tool used for Sabermetrics. Starting in 2006, PITCHf/x has recorded pitch-level data for individual players. The earlier years of the application are not as complete, with many missing values or even entire columns. ``pybaseball`` is a way to extract to PITCHf/x and other baseballdata directly into Python. You can read about the functions of the package [here](https://www.pydoc.io/pypi/pybaseball-1.0.1/). There are many R packages, such as ``baseballr``, ``Lahman``, and ``pitchRx`` that do a better job, but, as this is a Python class, I opted for the slightly less robust ``pybaseball``. 

Here is a way to install the package, if you wish:

In [ ]:
# !pip install pybaseball

In [40]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
import pybaseball

In [39]:
player_id = pybaseball.playerid_lookup('Verlander', first = 'Justin')
jv_id = player_id['key_mlbam'][0]
jv = pybaseball.statcast_pitcher(start_dt = '2017-04-02', end_dt = '2017-11-01', player_id = jv_id)
jv.head(10)

Gathering player lookup table. This may take a moment.
Gathering Player Data


pitch_type   game_date  release_speed  release_pos_x  release_pos_z  \
0         FF  2017-10-31           97.4        -1.9446         6.7237   
1         FF  2017-10-31           97.1        -1.8633         6.7674   
2         FF  2017-10-31           96.1        -1.8595         6.8649   
3         FF  2017-10-31           96.6        -1.8450         6.8482   
4         FC  2017-10-31           91.5        -2.2148         6.5903   
5         FC  2017-10-31           91.9        -2.1884         6.5650   
6         FC  2017-10-31           91.8        -2.3221         6.5049   
7         FC  2017-10-31           90.7        -2.2522         6.7167   
8         SL  2017-10-31           87.5        -2.3266         6.4741   
9         FF  2017-10-31           97.1        -1.9415         6.6737   

        player_name  batter  pitcher     events          description  \
0  Justin Verlander  641355   434378  strikeout      swinging_strike   
1  Justin Verlander  641355   434378        NaN                 foul   
2  Justin Verlander  641355   434378        NaN      swinging_strike   
3  Justin Verlander  641355   434378        NaN                 ball   
4  Justin Verlander  457759   434378  field_out        hit_into_play   
5  Justin Verlander  457759   434378        NaN                 foul   
6  Justin Verlander  457759   434378        NaN             foul_tip   
7  Justin Verlander  457759   434378        NaN                 foul   
8  Justin Verlander  608369   434378    sac_fly  hit_into_play_score   
9  Justin Verlander  608369   434378        NaN                 foul   

           ...           home_score  away_score  bat_score  fld_score  \
0          ...                    2           1          2          1   
1          ...                    2           1          2          1   
2          ...                    2           1          2          1   
3          ...                    2           1          2          1   
4          ...                    2           1          2          1   
5          ...                    2           1          2          1   
6          ...                    2           1          2          1   
7          ...                    2           1          2          1   
8          ...                    1           1          1          1   
9          ...                    1           1          1          1   

   post_away_score post_home_score post_bat_score post_fld_score  \
0                1               2              2              1   
1                1               2              2              1   
2                1               2              2              1   
3                1               2              2              1   
4                1               2              2              1   
5                1               2              2              1   
6                1               2              2              1   
7                1               2              2              1   
8                1               1              1              1   
9                1               1              1              1   

  if_fielding_alignment of_fielding_alignment  
0              Standard              Standard  
1              Standard              Standard  
2              Standard              Standard  
3              Standard              Standard  
4             Strategic              Standard  
5             Strategic              Standard  
6             Strategic              Standard  
7             Strategic              Standard  
8             Strategic              Standard  
9             Strategic              Standard  

[10 rows x 89 columns]

Here is a sample of the data. It is clearly loaded with (89!) variables to analyze. Perhaps it is better to view the overall structure of the dataframe.

In [43]:
jv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4065 entries, 0 to 4064
Data columns (total 89 columns):
pitch_type                         4065 non-null object
game_date                          4065 non-null object
release_speed                      4065 non-null float64
release_pos_x                      4065 non-null float64
release_pos_z                      4065 non-null float64
player_name                        4065 non-null object
batter                             4065 non-null int64
pitcher                            4065 non-null int64
events                             983 non-null object
description                        4065 non-null object
spin_dir                           0 non-null float64
spin_rate_deprecated               0 non-null float64
break_angle_deprecated             0 non-null float64
break_length_deprecated            0 non-null float64
zone                               4065 non-null int64
des                                983 non-null object
game_ty

Clearly the previous view of the dataframe did not do it justice. These are all of the variables available for all  4065 pitches JV threw in 2017. Exracting this data directly from [Baseball Savant](https://baseballsavant.mlb.com/statcast_search) or [Brook's Baseball](http://www.brooksbaseball.net/pfxVB/pfx.php?) would be a very tedious task, but ``pybaseball`` makes the process very quick and painless. There are some columns that contain all NULL values; however, it appears that this is because they were deprecated and are captured in a different variable.
## EDA with ``pandas``, ``matplotlib``, and ``seaborn``  
The principal function of ``pandas`` is the group by analysis, similiar to ``dplyr`` in R. This allows for interesting views of the data that can lead to conclusions that cannot be drawn from simply looking at the individual observations. An possible group by would be by ``game_date``, essentially giving the stats for each of JV's starts or pitching appearances. For this assignment I will be plotting using the ``seaborn`` API. This is mainly because we did not cover this topic in PCDA and I prefer the cleaner, more modern plots created from ``seaborn``.  

Let's do a group by ``game_date``:

In [48]:
jv.groupby(['game_date']).mean()

release_speed  release_pos_x  release_pos_z         batter  \
game_date                                                                
2017-04-04      90.933010      -2.520919       6.281090  551432.524272   
2017-04-10      89.208929      -2.489267       6.363852  524848.892857   
2017-04-15      91.604706      -2.410621       6.344404  526465.352941   
2017-04-21      91.063551      -2.452093       6.519348  553271.308411   
2017-04-27      91.637815      -2.489193       6.286086  512402.848739   
2017-05-02      91.275424      -2.431496       6.285583  527200.906780   
2017-05-09      91.434259      -2.337607       6.309935  542967.935185   
2017-05-14      91.477358      -2.422964       6.366220  513454.500000   
2017-05-20      91.731193      -2.407803       6.318866  535007.293578   
2017-05-25      91.249038      -2.339340       6.346934  548221.403846   
2017-05-30      91.430172      -2.471385       6.397753  515898.758621   
2017-06-04      91.313889      -2.549060       6.318433  556042.097222   
2017-06-10      90.205556      -2.477592       6.625239  556808.925926   
2017-06-15      89.681081      -2.523095       6.356530  526860.666667   
2017-06-21      90.875455      -2.316678       6.358909  526606.754545   
2017-06-27      89.782883      -2.384695       6.391414  518447.504505   
2017-07-02      90.815625      -2.565694       6.371705  517702.916667   
2017-07-08      91.928696      -2.427043       6.163005  526625.547826   
2017-07-14      90.903509      -2.556010       6.227872  474961.912281   
2017-07-19      91.773832      -2.359533       6.475934  502798.261682   
2017-07-24      91.112605      -2.313522       6.404791  513877.428571   
2017-07-30      91.770000      -2.309712       6.367613  551826.418182   
2017-08-04      91.243925      -2.183557       6.420978  518236.280374   
2017-08-09      90.421239      -2.357307       6.363026  530714.451327   
2017-08-15      88.995614      -2.529821       6.446669  457520.605263   
2017-08-20      92.682143      -2.313104       6.403169  511646.303571   
2017-08-25      92.212500      -2.203450       6.466046  587898.692308   
2017-08-30      91.507143      -2.389330       6.492726  503765.642857   
2017-09-05      92.129126      -2.256071       6.352665  525948.126214   
2017-09-12      91.665179      -2.099489       6.537375  491211.187500   
2017-09-17      89.792929      -2.164954       6.494784  533623.969697   
2017-09-22      90.882828      -2.169958       6.414408  494714.151515   
2017-09-27      90.176364      -2.439973       6.573657  561322.681818   
2017-10-05      92.243434      -2.206131       6.481120  573581.080808   
2017-10-09      92.107500      -2.270933       6.476000  588798.525000   
2017-10-14      91.764516      -2.193646       6.431826  527372.532258   
2017-10-20      91.108081      -2.146712       6.485989  539187.565657   
2017-10-25      92.858228      -2.081643       6.614192  578434.075949   
2017-10-31      93.721505      -2.018339       6.693434  566049.967742   

             pitcher  spin_dir  spin_rate_deprecated  break_angle_deprecated  \
game_date                                                                      
2017-04-04  434378.0       NaN                   NaN                     NaN   
2017-04-10  434378.0       NaN                   NaN                     NaN   
2017-04-15  434378.0       NaN                   NaN                     NaN   
2017-04-21  434378.0       NaN                   NaN                     NaN   
2017-04-27  434378.0       NaN                   NaN                     NaN   
2017-05-02  434378.0       NaN                   NaN                     NaN   
2017-05-09  434378.0       NaN                   NaN                     NaN   
2017-05-14  434378.0       NaN                   NaN                     NaN   
2017-05-20  434378.0       NaN                   NaN                     NaN   
2017-05-25  434378.0       NaN                   NaN                     NaN   
2017-05-30  43437